### PR TITLE
Improve performance on deeply nested topics

### PR DESCRIPTION
**Pull request by riedgar-ms**
Address report of poor performance from FC people, who were using deeply nested topics.

There are two main improvements:

- Simplify the subtopic checker so that it doesn't use a regular expression
- Rewrite `DataFrame.iterrows()` loops to use `DataFrame.apply()`

Put in basic tests of the functionality as well.

### CODE REVIEW

**Comment by riedgar-ms** [comment]
@Harsha-Nori I think this is ready to merge now.

File: adatest/_test_tree.py [hunk 1]
```diff
@@ -273,6 +273,22 @@ def topic(self, topic):
...
+        return hdt_df.any()
+
+    def topic_has_subtopics(self, target_topic: str) -> bool:
+        """Check if a topic has subtopics."""
+        has_subtopics_df = self._tests.apply(
+            lambda row: row['topic']!=target_topic and is_subtopic(target_topic, row["topic"]),
+            axis=1
+        )
+        return has_subtopics_df.any()
+
```
Comment (slundberg) [inline adatest/_test_tree.py:291]: If we still have performance issues we could try vector AND operations here. But this seems good for now :)

**Review by slundberg — APPROVED** [review]
Looks good to me. Thanks!

File: adatest/_test_tree.py (hunk 1 above)
Comment (riedgar-ms) [reply adatest/_test_tree.py:291]: If we stick with a topic tree, I'd be more inclined to rewrite the internal datastructure to be a tree, and just use DataFrame for serialisation.

**Pull request merged**